docs: Adding config.json documentation

garethtdavies opened the pull request:
Change-type: patch
Connects-to: #859
Signed-off-by: Gareth Davies gareth@balena.io

lurch commented on pages/reference/OS/configuration.md:
```diff
@@ -0,0 +1,179 @@
+(string) A space-separated list of preferred DNS servers to use for name resolution.
+
+* When `dnsServers` is not defined, or empty, the default value of Google's DNS (8.8.8.8) is used.
+* When `dnsServers` is `"null"` (a string), the system will not alter the dnsServers to the default server as detailed above.
```
Does this mean there'd be no DNS resolution at all?

pdcastro replied:
No, DNS servers obtained via DHCP (and/or [mentioned in a NetworkManager connection profile](https://www.balena.io/docs/reference/OS/network/2.x/#setting-a-static-ip)?) are also used. Also, I think (hope!) that the following paragraph in the [Network Requirements](https://www.balena.io/docs/reference/OS/network/2.x/#network-requirements) docs is correct, and could be added / copied (?) to this section (or have it only here, and change the network requirements to link here...)

> Google's Public DNS server at 8.8.8.8 is used by default and in addition to DNS servers obtained via DHCP (balenaOS may issue queries to multiple DNS servers simultaneously, for the quickest response to be used). If additional DNS servers are configured via DHCP or other means, it is OK for the local network to block requests to 8.8.8.8. To avoid any requests being made to 8.8.8.8 by balenaOS, modify the dnsServers setting in config.json.

So maybe actually the three bullet points would be more informative if rewritten as follows, though this needs testing to confirm, or source code to be read, or expert advice from perhaps @majorz or @alexgg (Zubair is away for a while...)
* When `dnsServers` is not defined, or empty, Google's DNS server (8.8.8.8) is added to the list of DNS servers obtained via DHCP or statically configured in a NetworkManager connection profile.
* When `dnsServers` is `"null"` (a string), Google's DNS server (8.8.8.8) will NOT be added as described above.
* When `dnsServers` is defined and not `"null"`, the listed servers will _replace_ any servers obtained via DCHP or statically configured via a NetworkManager connection profile.

alexgg replied:
Hi @pdcastro, how I understand the DNS resolution in Balena is that it uses dnsmasq as a local cache and forwarder.
/etc/resolv.conf is symlinked to resolv.dnsmasq
and dnsmasq is started with:
`-r /etc/resolv.dnsmasq -z --servers-file=/run/dnsmasq.servers`
`/etc/resolv.dnsmasq` is a symlink to `/var/run/resolvconf/interface/NetworkManager`, which is the file that NetworkManager uses to store the DHCP or statically configured DNS server. This is the first place where dnsmasq lookes for upstream servers.
And `/run/dnsmasq.servers` is the file that is configured from config.json with either the user provided servers or 8.8.8.8.
So dnsServers only affects the contents of `/run/dnsmasq.servers`, and the NM DHCP or static configuration will always be there.
So I would rephase the above to:

- When `dnsServers` is not defined, or empty, Google's DNS server (8.8.8.8) is added to the list of DNS servers obtained via DHCP or statically configured in a NetworkManager connection profile.
- When `dnsServers` is "null" (a string), Google's DNS server (8.8.8.8) will NOT be added as described above.
- When `dnsServers` is defined and not "null", the listed servers will be added to the list of servers obtained via DCHP or statically configured via a NetworkManager connection profile.

pdcastro replied:
Thank you for digging this @alexgg! 👍 We are finally getting more clarity on these points. It surprises me that `--all-servers` isn't the default behaviour, as I believe this is the "support knowledge" that we share with users. @majorz, are you similarly surprised? :-) 
In this case, we need to update the Network Requirements page as well, as it is incorrect.

garethtdavies replied:
There is some additional material here on this issue just published via @hedss WIP masterclass https://hackmd.io/7z1NBeDoTy2rqchIC9ogHg?view#52-DNS-Issues

garethtdavies replied:
Opened an issue to track this https://github.com/balena-io/docs/issues/1283

pdcastro commented on pages/reference/OS/configuration.md:
````diff
@@ -0,0 +1,179 @@
+}
+```
+
+## Valid fields
````
This section looks a lot like: https://github.com/balena-os/meta-balena#configjson
So there's a source of truth / sync issue: when someone merges a meta-balena PR that modifies the README, how does it get (automatically) propagated to the docs repo? We have an existing setup with scripts that copy md files from other repos - cli, supervisor, Node SDK... Should we setup something like that? Maybe the config.json docs could be moved out of the meta-balena README and into its own separate md file - also in the meta-balena repo?

garethtdavies replied:
Yes, this is largely taken verbatim with a few fixes and that's a very good idea. I was thinking along those same lines by importing them as a partial. Let's complete this document in this repo for now (otherwise trying to coordinate reviews in two repos would be painful) and then once it is complete I'll move this section, to a new page in the meta-balena repo.

garethtdavies replied:
This will be addressed in a followup PR. I've extracted this to a partial and from the PR will be generated from the meta-balena repo.

pdcastro commented on pages/reference/OS/configuration.md:
```diff
@@ -0,0 +1,179 @@
+
+Alternatively, you can always reprovision a device with an updated `config.json` file.
+
+**Note:** Editing config.json on a provisioned device is not recommended, as mistakes to this file can leave a device inaccessible. If you need to manually edit this file, see the [balenaOS masterclass instructions](https://github.com/balena-io-projects/balenaos-masterclass#12-advanced-editing-configjson-by-hand).
```
This paragraph partially duplicates another paragraph above _("Special care should be taken...")_ Might be nicer to merge the 2 paragraphs, and perhaps emphasise a whole warning sentence in bold (if the reason to  duplicate the information was to make sure it is noticed...).

garethtdavies replied:
Agreed. I'm also not sure whether we should be recommending this or not - the masterclass also has its own fair share of warnings if we do link to it :)

pdcastro replied:
> I'm also not sure whether we should be recommending this or not 

**My two cents,** I think we should, yes, because:
* The masterclasses are public and were written with the intent of sharing with end users. We share them in support interactions as well.
* In forums or paid support, users often ask about editing `config.json` and we share this kind of advice anyway, though perhaps inconsistently between support agents. Having the information in a public documentation page support effort and improves advice consistency. And if users try it and it goes wrong in some step, we will know where to focus the improvement effort.

I wonder if @shaunmulligan thinks the same way... :-)

hedss replied:
I agree we should share this advice, but we definitely want to really stress the point that if they make changes that makes the file unreadable, they're going to be in a position where they won't be able to access the device again. Saying that, if it's already provisioned and had connected to the VPN *and* there's another device online on the same network, we could probably salvage it. But that's not a support ticket we want to see :)

hedss replied:
And actually, adding to this, if people do want to mess around with it, you want to do it in a safe way: https://github.com/balena-io-playground/configizer

pdcastro commented on pages/reference/OS/configuration.md:
```diff
@@ -0,0 +1,179 @@
+
+### ntpServers
+
+(string) A space-separated list of NTP servers to use for time synchronization. Defaults to `resinio.pool.ntp.org` servers.
```
The [Network Requirements](https://www.balena.io/docs/reference/OS/network/2.x/#network-requirements) docs have a full list of NTP servers:

> NTP / UDP packets (port 123) are exchanged with:
> 0.resinio.pool.ntp.org
> 1.resinio.pool.ntp.org
> 2.resinio.pool.ntp.org
> 3.resinio.pool.ntp.org

But the Network Requirements docs fail to mention that the servers could be modified via `config.json`. So perhaps we could list the 4 servers above here, and modify the Network Requirements docs to point here for a list of servers.

pdcastro commented on pages/reference/OS/configuration.md:
```diff
@@ -0,0 +1,179 @@
+
+### country
+
+(string) The country in which the device is operating. Used for setting the WiFi regulatory domain.
```
```
(string) Two-letter country code for the region the device is operating in.
Used for setting the WiFi regulatory domain.
```
I suspect not all country codes are supported and it's tempting to add something like _"Check the WiFi device driver for a list of supported country codes"_ and to have "Two-letter country code" be a link to [ISO 3166-1 alpha-2 on Wikipedia](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2), but I'm not sure it's correct... :-)

lurch replied:
In case it has useful / relevant information: https://github.com/raspberrypi/documentation/blob/master/configuration/wireless/wireless-cli.md

pdcastro commented on pages/reference/OS/configuration.md:
```diff
@@ -0,0 +1,179 @@
+* "uri" (string) Value of the url to query for connectivity checks. Defaults to `$API_ENDPOINT/connectivity-check`.
+* "interval" (string) Interval between connectivity checks in seconds. Defaults to 3600.
+* "response" (string). If set controls what body content is checked for when requesting the URI. If it is an empty value, the HTTP server is expected to answer with status code 204 or send no data.
+
```
Maybe we can add here something like:
> This feature builds on NetworkManager's connectivity check, which is further documented in the connectivity section at: https://developer.gnome.org/NetworkManager/stable/NetworkManager.conf.html

lurch replied:
Is it worth mentioning (or linking to other docs) what happens if the connectivity check fails?

garethtdavies replied:
I've left this with the link to https://developer.gnome.org/NetworkManager/stable/NetworkManager.conf.html for more in-depth discussion.

pdcastro commented on pages/reference/OS/configuration.md:
```diff
@@ -0,0 +1,179 @@
+
+## Using config.json to provision a device
+
+The `config.json` file is located in the root of the `resin-boot` partition, and you may edit this file by mounting a flashed SD card and editing the file directly. Editing this file will only work if you edit before the device's first boot. Once the device has been provisioned, edits to this file have no effect as the contents of the file are copied to the device boot partition at `/mnt/boot/con ...
```
@shaunmulligan or @alexgg might be able to help on this point as well...

hedss replied:
Yep, you can edit the `config.json` on-device, `/mnt/boot` is a mapping to the boot partition (first) of the media. We shouldn't really be mentioning `/resin-boot`, these are backups to ensure that should a HUP fail. Weirdly, as I started typing here, this thread: https://www.flowdock.com/app/rulemotion/user_happiness/threads/o2DA9mNhyUEafWsSUMauNi0t3se appeared. Worth reading. :)

garethtdavies commented:
OK, thanks, everyone. Updated this file to incorporate everything above. Apologies, I squashed so you can't see review changes but enough was changed to probably warrant a re-scan.

I've also done a lot more testing on this, not least as that `/resin-boot` directory threw me off. I was also confused as the supervisor will only overwrite settings on a provisioned device it manages - so you can't say change persistent logging on a provisioned device but you can add SSH keys / udev rules (basically anything that configizer allows). This issue is also interesting to make this behavior consistent https://github.com/balena-io/balena-supervisor/issues/998. 

As noted above, I'll arrange the content to avoid duplicating once the text is OK.

pdcastro commented on pages/reference/OS/configuration.md:
```diff
@@ -0,0 +1,193 @@
+
+There are various ways of safely modifying the values of `config.json` on a provisioned device:
+
+- Using the supervisor API, you can update the device [hostname][hostname].
```
```
Using the supervisor API, you can update the device [hostname and proxy settings][hostconfig]

[hostconfig]: /reference/supervisor/supervisor-api/#patch-v1devicehost-config
```

pdcastro commented on pages/reference/OS/configuration.md:
```diff
@@ -0,0 +1,193 @@
+
+The behavior of {{ $names.os.lower }} can be configured by editing the `config.json` file. This file is located in the [boot partition][boot-partition] and accepts a [range of fields](#valid-fields) to modify the behavior of the host OS.
+
+Before the device has been provisioned, you may edit `config.json` by mounting a flashed SD card (which will mount as `resin-boot`) and editing the file directly. The boot partition is mounted on the device at `/mnt/boot`, and so on the device, the file is located at `/mnt/boot/config.json`. For example, to output a formatted version of `config.json` on a device, use the following commands:
```
* has been provisioned → is provisioned
* This paragraph describes how to edit the `config.json` file ***before*** the device is provisioned. Then there is a whole section **"Sample config.json",** a long section **"Valid fields",** and then the section **"Editing config.json on a provisioned device".** I think this page would be better organised by having the complementary options of editing `config.json` _before and after_ provisioning in the same section, or in consecutive sections, rather than in "distant" sections.
* I am not sure how obvious the meaning of "provisioned" is. I suggest making the word "provisioned" a link to https://www.balena.io/docs/learn/welcome/primer/#device-provisioning

garethtdavies replied:
Thanks again for the feedback. The distinction between before and after provisioning is _really_ confusing as the behavior is inconsistent. I did a lot of testing on this and determined the following behavior:

* Before or after provisioning you can edit _some_ values of config.json so long as the Supervisor doesn't manage them. I mounted an SD card from a provisioned device into my laptop and added the SSH keys or udev rules, for example, and that works just fine.
* When you edit `persistentLogging` before the device is provisioned this would be respected but if you edit this _after_ the device has been provisioned the Supervisor overwrites this setting - it actually reformats the file too (to minify the json). I suspect the same would be true of the hostname too (I'll test that later)

So the distinction between before and after provisioning is very muddled and has caveats, but I may just make this a single section and mention these.

pdcastro replied:
@garethtdavies, what you've described above matches my previous "anecdotal" experience -- but I had not conducted an experiment (or read source code...) to allow me to state it. It really sounds correct though, and actually perhaps you could add a sentence in the lines of:
_"After provisioning, editing the `config.json` as described above is not reliable or advisable because the supervisor may overwrite some fields, such as persistentLogging, with values read from the balenaCloud (API) database."_
Just a thought. :-)

garethtdavies replied:
I think we are getting there.... 😆 This behavior is also confirmed by configizer which only allows specifying of fields not managed by the Supervisor.

Thanks for that suggestion. I think something like that will work and am reformating this now. I don't want to blanket advise against modifying it (with a suitable warning) as for example, adding an SSH key for a prod image is incredibly useful. Those changes _do_ work consistently and I could imagine being popular requests.

garethtdavies replied:
I pushed an update and reorganized so it flows a bit better - i.e. the how to safely edit (three different ways depending on what you want to do...) follows the advice to not do this directly on a provisioned device.

lurch replied:
Is there a reliable list of which fields can be edited in `config.json` before provisioning (all of them?), and which fields _shouldn't_ be edited after provisioning, because the supervisor will overwrite any changes? Or is that wildly dependent on the particular version of the supervisor?

pdcastro replied:
> Is there a reliable list of which fields can be edited in `config.json`

I may be misreading the supervisor source code, but it would appear that any of the fields listed in this [schema.ts](https://github.com/balena-io/balena-supervisor/blob/v10.6.26/src/config/schema.ts) file (like `uuid`, `apiKey`, `hostname`, `persistentLogging`...) are managed by the supervisor, and thus cannot be edited manually. Then this block of code seems to read values from a database and update `config.json`:
https://github.com/balena-io/balena-supervisor/blob/v10.6.26/src/config/index.ts#L153-L180

But I am not 100% sure, and perhaps @CameronDiver could comment further. Also, I don't know if it is a good idea to publish such a list, because this might prevent the supervisor from managing the fields in the future. Unless the list was explicitly linked to supervisor versions, alongside a warning that updating the host OS might break the user's app or scripts if they are "manually" editing some fields...

garethtdavies replied:
I assumed the list here (valid fields) was the list of editable fields. 

My understanding then is after provisioning hostname (supervisor API) and persistent logging  (config variable) and the rest we recommend configizer (which only allows these other fields). I guess this is why we don't want users to manually edit this file as it can be unpredictable as noted in the text :)

lurch commented on pages/reference/OS/configuration.md:
```diff
@@ -0,0 +1,193 @@
+
+The behavior of {{ $names.os.lower }} can be configured by editing the `config.json` file. This file is located in the [boot partition][boot-partition] and accepts a [range of fields](#valid-fields) to modify the behavior of the host OS.
+
+**Note:** The `config.json` file is different from the [`config.txt`][config-txt] file, also located in the boot partition, which is used by the Raspberry Pi to set device configuration options.
```
I wonder if it's worth mentioning "...which is (only) used by the Raspberry Pi to...", to highlight the fact that non-RaspberryPi devices don't use `config.txt`? :man_shrugging:

garethtdavies replied:
Thanks for the feedback. I decided to leave this one out as I think it is mostly redundant given the context here is more to draw attention to the fact a similar filename exists.

lurch commented on pages/reference/OS/configuration.md:
```diff
@@ -0,0 +1,193 @@
+
+**Note:** The `config.json` file is different from the [`config.txt`][config-txt] file, also located in the boot partition, which is used by the Raspberry Pi to set device configuration options.
+
+Before the device is [provisioned][provisioned], you may edit `config.json` by mounting a flashed SD card (which will mount as `resin-boot`) and editing the file directly. The boot partition is mounted on the device at `/mnt/boot`, and so on the device, the file is located at `/mnt/boot/config.json`. For example, to output a formatted version of `config.json` on a device, use the following comman ...
```
I wonder if "(with the partition-label `resin-boot`)" is better than "(which will mount as `resin-boot`)"; since I guess the latter is OS-specific? (e.g. think of Windows users)  :man_shrugging:

lurch commented on pages/reference/OS/configuration.md:
````diff
@@ -0,0 +1,193 @@
+cat /mnt/boot/config.json | jq '.'
+```
+
+**Warning:** Editing config.json on a provisioned device should be done very carefully as any mistakes to this file can leave a device inaccessible.
````
nit: "mistakes to this file" doesn't sound right. Perhaps "any mistakes made in the syntax of this file" or "any errors created in this file" ?

garethtdavies commented:
Thanks all, addressed the above. There will be a quick follow-up PR to move the new valid fields partial to the meta-balena repo as this will be pulled in dynamically.

I've opened an issue for the potential issue for Network manager docs (dnsmasq behavior) https://github.com/balena-io/docs/issues/1283

garethtdavies commented:
@balena-ci retest

garethtdavies commented:
@balena-cli retest

hedss commented on pages/reference/OS/configuration.md:
```diff
@@ -0,0 +1,76 @@
+[configuration]: /learn/manage/configuration/
+[nm-connectivity]: https://developer.gnome.org/NetworkManager/stable/NetworkManager.conf.html
+[configizer]:{{ $links.githubPlayground }}/configizer
+[masterclass-os]:{{ $links.githubProjects }}/balenaos-masterclass#12-advanced-editing-configjson-by-hand
```
I've moved these all over to the `balena-io` org now.

hedss commented on shared/general/config-json-fields.md:
```diff
@@ -0,0 +1,118 @@
+
+(string) A space-separated list of preferred DNS servers to use for name resolution.
+
+- When `dnsServers` is not defined, or empty, Google's DNS server (8.8.8.8) is added to the list of DNS servers obtained via DHCP or statically configured in a NetworkManager connection profile.
```
It actually uses both the Google primary and secondary servers (`8.8.8.8` and `8.8.4.4`), see a `/etc/resolv.dnsmasq` config file on-device.

garethtdavies replied:
Just checking this as this info came from @alexgg comment above https://github.com/balena-io/docs/pull/1249#discussion_r354797800. Specifically `And /run/dnsmasq.servers is the file that is configured from config.json with either the user provided servers or 8.8.8.8.` 

On my sample device, this appears to be true and `8.8.4.4` is not added.

```
root@3a628fc:~# cat /run/dnsmasq.servers

server=8.8.8.8
```

hedss replied:
OK, that's interesting, maybe later OS changes have changed this (I looked on a 2.38).

garethtdavies replied:
I think it's this line? https://github.com/balena-os/meta-balena/blob/a7c5f47ef43c9ef79b770ec6ac83989130f5f4ed/meta-balena-common/recipes-support/resin-vars/resin-vars/resin-vars#L90

garethtdavies commented:
@balena-ci retest

garethtdavies commented:
@balena-ci retest
